### PR TITLE
feat(forms): new API for working with sets of fields

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds input/fieldset view with a alignment and justification options
Adds elgg_view_field() as an alias to elgg_view_input() to render
fields from an associative array.
Improves dev usability of buttons.
Updates discussions plugin to new API.
- [ ] Update upgrading docs

fieldset

### CODE REVIEW

File: engine/lib/views.php [hunk 1]
```diff
@@ -1370,13 +1370,13 @@ function elgg_get_form_footer() {
...
  * Renders a form field
  *
- * @param string $input_type Input type, used to generate an input view ("input/$input_type")
- * @param array  $vars       Fields and input vars.
- *                           Field vars contain both field and input params. 'label', 'help',
- *                           and 'field_class' params will not be passed on to the input view.
- *                           Others, including 'required' and 'id', will be available to the
- *                           input view. Both 'label' and 'help' params accept HTML, and
- *                           will be printed unescaped within their wrapper element.
+ * @param mixed $input_type Input type, used to generate an input view ("input/$input_type")
```
Comment (hypeJunction) [inline engine/lib/views.php]: Scratch that

File: engine/lib/views.php [hunk 2]
```diff
@@ -1428,6 +1429,25 @@ function elgg_view_input($input_type, array $vars = array()) {
...
+ * offers support for:
+ *  - #type: input type
+ *
+ * @see elgg_view_input()
+ *
+ * @param array $vars Field vars
+ * @return string
+ */
+function elgg_view_field(array $vars = []) {
+	$input_type = elgg_extract('#type', $vars);
```
Comment (hypeJunction) [inline engine/lib/views.php]: I am thinking in the future, we will follow this Drupaley approach to decorating fields with validation rules, various setter/getter/sanitizer callbacks.

File: mod/aalborg_theme/views/default/elements/buttons.css [hunk 3]
```diff
@@ -18,6 +18,12 @@
 .elgg-button:focus {
 	text-decoration: none;
 }
+.elgg-button .elgg-icon {
+    color: inherit;
```
Comment (hypeJunction) [inline mod/aalborg_theme/views/default/elements/buttons.css]: I just really hate that I have to do this all the time. Do we really need to set color on the icons?

File: mod/developers/views/default/theme_sandbox/forms.php [hunk 4]
```diff
@@ -189,6 +189,66 @@
 			'label' => 'Number input (.elgg-input-number) with custom options:',
 			'help' => 'Enter an integer number larger than zero',
 		));
+
+		$dt = new \DateTime(null, new \DateTimeZone('UTC'));
+		$hour_options = array();
+		$hour_options_ts = range(0, 24*60*60, 900); // step of 15 minutes
+		foreach ($hour_options_ts as $ts) {
+			$hour_options[$ts] = $dt->setTimestamp($ts)->format('g:ia');
```
Comment (hypeJunction) [inline mod/developers/views/default/theme_sandbox/forms.php]: Do we want `input/time` with configurable minute step in core?

File: views/default/input/form.php [hunk 5]
```diff
@@ -23,17 +25,40 @@
...
+	], $vars['header']);
+	unset($vars['header']);
+}
+
+$body = elgg_format_element('fieldset', [
+	'class' => 'elgg-form-body',
+], $tokens . $body);
+
+$footer = '';
+if (isset($vars['footer'])) {
```
Comment (hypeJunction) [inline views/default/input/form.php]: I initially thought I would make the PR against 2.x, but it's not going to work. This is where the deferred footer should go.

File: mod/developers/views/default/theme_sandbox/forms.php (hunk 4 above)
Comment (jdalsem) [inline mod/developers/views/default/theme_sandbox/forms.php]: would be nice

File: views/default/input/button.php [hunk 6]
```diff
@@ -16,19 +17,21 @@
...
+		$vars['class'][] = 'elgg-button-submit';
+		break;
+	case 'reset':
+		$vars['class'][] = 'elgg-button-cancel';
+		break;
+	case 'button':
 	case 'image':
 		break;
 	default:
 		$vars['type'] = 'button';
```
Comment (jdalsem) [inline views/default/input/button.php]: is this still needed if the element is already a button??

File: views/default/input/form.php [hunk 7]
```diff
@@ -23,17 +25,40 @@
 $vars = array_merge($defaults, $vars);
 
 $vars['class'] = elgg_extract_class($vars, 'elgg-form');
-$vars['action'] = elgg_normalize_url($vars['action']);
-$vars['method'] = strtolower($vars['method']);
 
-$body = $vars['body'];
+$vars['action'] = elgg_normalize_url(elgg_extract('action', $vars, ''));
+unset($vars['action_name']);
```
Comment (jdalsem) [inline views/default/input/form.php]: why unsetting the action_name here? it is not used before... shouldn't it be removed at all?

File: views/default/input/submit.php [hunk 8]
```diff
@@ -6,10 +6,16 @@
...
+ * @uses $vars['value'] Value of the submit input
  */
 
 $vars['type'] = 'submit';
 
 $vars['class'] = elgg_extract_class($vars, 'elgg-button-submit');
 
+if (!isset($vars['text']) && isset($vars['value'])) {
+	$vars['text'] = $vars['value'];
+}
```
Comment (jdalsem) [inline views/default/input/submit.php]: is there a need to unset value here?

File: views/default/input/form.php (hunk 7 above)
Comment (hypeJunction) [inline views/default/input/form.php]: It comes from `elgg_view_form()`, probably some legacy code

File: views/default/input/submit.php (hunk 8 above)
Comment (hypeJunction) [inline views/default/input/submit.php]: No, you want the `value` attribute of the button if it's set and someone passes `name` as well.

File: engine/lib/views.php (hunk 2 above)
Comment (mrclay) [inline engine/lib/views.php]: :+1:

File: mod/aalborg_theme/views/default/elements/buttons.css (hunk 3 above)
Comment (mrclay) [inline mod/aalborg_theme/views/default/elements/buttons.css]: I'm open to suggestion, particularly for 3

**Comment by mrclay** [comment]
Let's teach elgg_view_input to handle type "markup" too.

File: mod/aalborg_theme/views/default/elements/buttons.css (hunk 3 above)
Comment (hypeJunction) [inline mod/aalborg_theme/views/default/elements/buttons.css]: Just remove the color globally and only set it where appropriate

**Comment by hypeJunction** [comment]
I going to break this down into 2 PRs, features for 2.x and breaks for 3.x

**Comment by hypeJunction** [comment]
First replacement PR #10331

**Comment by hypeJunction** [comment]
This has been broken down into multiple PRs

**Pull request closed without merging**